### PR TITLE
expression: add interfaces for refactoring typeinferer

### PR DESCRIPTION
**Pull request by XuHuaiyu**
Before this commit, functions in TiDB resolves types during the calculation，
this commit resolve types in plan-build phrase, in order to optimize the calculating process of functions.

This commit implements the new interface of type resolve.

refer to doc: https://docs.google.com/document/d/1YuftfsNJTmplhwNAnSgeYVk-ZCCyBrEOaCorL593yNI/edit#heading=h.3i90ec2a7uma
PTAL @shenli @coocood @zimulala @hanfei1991

### CODE REVIEW

File: expression/builtin.go [hunk 1]
```diff
@@ -23,6 +23,7 @@ import (
 	"github.com/pingcap/tidb/context"
 	"github.com/pingcap/tidb/parser/opcode"
 	"github.com/pingcap/tidb/util/types"
+	"strconv"
```
Comment (coocood) [inline expression/builtin.go]: fix import format

File: expression/builtin.go [hunk 2]
```diff
@@ -83,10 +121,209 @@ func (b *baseBuiltinFunc) getCtx() context.Context {
...
+	err = res.FromString([]byte(val))
+	return res, false, errors.Trace(err)
+}
+
 // builtinFunc stands for a particular function signature.
 type builtinFunc interface {
 	// eval does evaluation by the given row.
 	eval([]types.Datum) (types.Datum, error)
+	// evalInt evaluate int result of builtinFunc by given row.
+	evalInt(row []types.Datum) (int64, bool, error)
```
Comment (coocood) [inline expression/builtin.go]: The returned values should be commented or named like
`evalInt(row []types.Datum) (val int64, isNull bool, err error)

File: expression/builtin.go [hunk 3]
```diff
@@ -83,10 +121,209 @@ func (b *baseBuiltinFunc) getCtx() context.Context {
 	return b.ctx
 }
 
+// baseIntBuiltinFunc represents the functions which return int values.
+type baseIntBuiltinFunc struct {
+	baseBuiltinFunc
+}
+
+func (b *baseIntBuiltinFunc) eval(row []types.Datum) (d types.Datum, err error) {
+	res, isKindNull, err := b.self.evalInt(row)
```
Comment (coocood) [inline expression/builtin.go]: just `isNull` will do.

**Comment by coocood** [comment]
LGTM

**Comment by coocood** [comment]
@hanfei1991 PTAL

File: expression/builtin.go [hunk 4]
```diff
@@ -83,10 +122,209 @@ func (b *baseBuiltinFunc) getCtx() context.Context {
...
 }
 
+// baseIntBuiltinFunc represents the functions which return int values.
+type baseIntBuiltinFunc struct {
+	baseBuiltinFunc
+}
+
+func (b *baseIntBuiltinFunc) eval(row []types.Datum) (d types.Datum, err error) {
+	res, isNull, err := b.self.evalInt(row)
+	if err != nil {
```
Comment (shenli) [inline expression/builtin.go]: if err != nil || isNull {
Keep consistent with the following code.

File: expression/expression.go [hunk 5]
```diff
@@ -45,13 +46,99 @@ const (
 // EvalAstExpr evaluates ast expression directly.
 var EvalAstExpr func(expr ast.ExprNode, ctx context.Context) (types.Datum, error)
 
+type baseExpr struct {
```
Comment (shenli) [inline expression/expression.go]: Add comment for this structure.

**Comment by XuHuaiyu** [comment]
@shenli  PTAL

**Comment by shenli** [comment]
LGTM
@XuHuaiyu Please resolve the conflicts.

**Pull request merged**